### PR TITLE
CONTRACTS: loop assigns clause inference function

### PR DESCRIPTION
**Pull request by remi-delmas-3000**
The first commit is from #7636, do not review.

Co-authored with @qinheping

Loop assigns clause inference function. Uses `loop_utils.h` to get a set of lvalue expressions modulo aliasing that may be assigned by the loop. 
The lvalue expressions are then post processed. 

- If an inferred target depends on a loop-local variable (i.e. a variable that does not exist outside of the loop body) it is widened to its base object. If the base object still depends on a loop local, the target is discarded, otherwise `__CPROVER_object_whole(&root_object)` is kept as inferred target.
- If the address of an inferred target is non-constant, then `__CPROVER_object_whole(&target)` is used as widened target.

Extracted from https://github.com/diffblue/cbmc/pull/7541, can only be tested once all features are in place.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.h
Comment (tautschnig) [inline src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.h]: Is a `set` the most appropriate data structure here? Do you actually need the ordering? I'd otherwise suggest to use `std::unordered_set`.

File: src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.h
Comment (tautschnig) [inline src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.h]: Should be `const`

File: src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.cpp
Comment (tautschnig) [inline src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.cpp]: Can we change `get_assigns` so that it returns by value?

File: src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.cpp
Comment (tautschnig) [inline src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.cpp]: Probably `std::unordered_set`

File: src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.cpp
Comment (tautschnig) [inline src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.cpp]: Is it possible for `*expr_it` to be, e.g., `A[i]`? If so, wouldn't `i` wrongly end up in this set? Don't you actually want what is being built in #7626?

File: src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.cpp
Comment (tautschnig) [inline src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.cpp]: Is it correct to remove an expression from `assigns` if _any_ symbol occurring in there is a loop-local one? (At least that is my understanding of what the `find_if` will accomplish here.)

File: src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.cpp
Comment (tautschnig) [inline src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.cpp]: Which of the above code can actually throw? It seems it could only be `get_assigns`, but I wasn't able to find a `throw` over there either.

File: src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.cpp
Comment (tautschnig) [inline src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.cpp]: `const code_typet &`

File: src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.cpp
Comment (tautschnig) [inline src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.cpp]: No source location is set for `address_of_expr`, so this will always be `nil`.

File: src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.cpp
Comment (tautschnig) [inline src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.cpp]: The same `symbol_expr` will be constructed repeatedly.

File: src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.cpp
Comment (tautschnig) [inline src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.cpp]: `exprt::operandst args{{address_of_expr}};` should facilitate RAII.

**Review by tautschnig — CHANGES_REQUESTED** [review]

File: src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.h
Comment (remi-delmas-3000) [reply src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.h]: The order is not relevant, but this is due to `loop_util.h` returning `assignst` defined as `typedef std::set<exprt> assignst`. 

I'd rather change the return type to `assignst` instead of adding a translation from `set` to `unordered_set`. Would that be OK ?

File: src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.h
Comment (remi-delmas-3000) [reply src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.h]: fixed

File: src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.cpp
Comment (qinheping) [reply src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.cpp]: I removed the `throw` in #7603 as even we failed to compute alias assigns, it is still meaningful to return what we infer.

File: src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.cpp
Comment (remi-delmas-3000) [reply src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.cpp]: `loop_utils.h` defines `get_assigns` like this, and the whole k-induction module depends on this interface so I'd rather not change it.

File: src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.cpp
Comment (remi-delmas-3000) [reply src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.cpp]: changed to `assignst`

File: src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.cpp
Comment (remi-delmas-3000) [reply src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.cpp]: Removing something from assigns is always sound as it can only cause the analysis to fail with an assigns clause violation.

We really have to remove any inferred expression that contains a non-local symbol, because assigns clause targets can only talk about variables that exist outside of the loop.

```
for (int i = 0; i < N; i++) {
 int j = i;
 arr[j] = 0;
}
```
In that case the inference returns the expressions `i` and `arr[j]` as inferred targets, and we cannot use `arr[j]`.

Maybe we could do the following instead: if a target depends on some loop local, compute `root = root_object(inferred_target)`,  if `root` still contains non-local variables, discard it, otherwise widen to `__CPROVER_object_whole(root)` and add it to the set.
 
  @qinheping does that sound right ?

File: src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.cpp
Comment (remi-delmas-3000) [reply src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.cpp]: so should we remove the try/catch completely ?

Comment (qinheping) [reply src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.cpp]: Yes, we should remove the try/catch.

Comment (remi-delmas-3000) [reply src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.cpp]: I removed the try-catch

File: src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.cpp
Comment (remi-delmas-3000) [reply src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.cpp]: fixed

File: src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.cpp
Comment (remi-delmas-3000) [reply src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.cpp]: fixed

File: src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.cpp
Comment (remi-delmas-3000) [reply src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.cpp]: How can I prevent that ?

File: src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.cpp
Comment (remi-delmas-3000) [reply src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.cpp]: fixed

**Comment by remi-delmas-3000** [comment]
@tautschnig @qinheping I just pushed a revised version that uses root objects to widen targets, should be stable now

File: src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.cpp
Comment (qinheping) [reply src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.cpp]: In the example, `get_assigns` will give us `j` and `arr[j]`. 
Yes, we should remove the local variable `j`.
For deference expressions with non-constant offset, e.g., `*(arr+j)`; and indexing expressions with non-constant index, e.g., `arr[j]`, we should widen them with `__CPROVER_object_whole`.

File: src/goto-instrument/contracts/dynamic-frames/dfcc_root_object.h
Comment (feliperodri) [inline src/goto-instrument/contracts/dynamic-frames/dfcc_root_object.h]:
```suggestion
/// Computes a set of root object expressions for an assignment `lhs` expression.
```

File: src/goto-instrument/contracts/dynamic-frames/dfcc_root_object.h
Comment (feliperodri) [inline src/goto-instrument/contracts/dynamic-frames/dfcc_root_object.h]:
```suggestion
/// is expected not to contain disjunctions. Fails if the expression contains
```

File: src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.cpp
Comment (feliperodri) [inline src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.cpp:37]: Could you also add a comment header here?

File: src/goto-instrument/contracts/dynamic-frames/dfcc_root_object.cpp
Comment (feliperodri) [inline src/goto-instrument/contracts/dynamic-frames/dfcc_root_object.cpp]:
```suggestion
  const exprt &lhs,
```

File: src/goto-instrument/contracts/dynamic-frames/dfcc_root_object.cpp
Comment (feliperodri) [inline src/goto-instrument/contracts/dynamic-frames/dfcc_root_object.cpp]:
```suggestion
/// Recursively computes a set of root object expressions for an assignment `lhs`.
```

File: src/goto-instrument/contracts/dynamic-frames/dfcc_root_object.cpp
Comment (feliperodri) [inline src/goto-instrument/contracts/dynamic-frames/dfcc_root_object.cpp]:
```suggestion
  const exprt &lhs,
```

File: src/goto-instrument/contracts/dynamic-frames/dfcc_root_object.cpp
Comment (feliperodri) [inline src/goto-instrument/contracts/dynamic-frames/dfcc_root_object.cpp]:
```suggestion
/// dereference expressions, so that `object_descriptor_exprt` can be used to
```

**Review by feliperodri — APPROVED** [review]

File: src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.cpp
Comment (remi-delmas-3000) [reply src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.cpp]: Yes if an expression is of the form `a[i]` with `i` non constant or a loop local, post processing is applied to widen the inferred target to the underlying root object. #7636 is now included as first commit in this PR.

File: src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.cpp
Comment (remi-delmas-3000) [reply src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.cpp]: This is what is implemented in this new revision, thanks to #7636

File: src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.cpp
Comment (remi-delmas-3000) [reply src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.cpp:37]: fixed

File: src/goto-instrument/contracts/dynamic-frames/dfcc_root_object.h
Comment (remi-delmas-3000) [reply src/goto-instrument/contracts/dynamic-frames/dfcc_root_object.h]: fixed

File: src/goto-instrument/contracts/dynamic-frames/dfcc_root_object.h
Comment (remi-delmas-3000) [reply src/goto-instrument/contracts/dynamic-frames/dfcc_root_object.h]: fixed

File: src/goto-instrument/contracts/dynamic-frames/dfcc_root_object.cpp
Comment (remi-delmas-3000) [reply src/goto-instrument/contracts/dynamic-frames/dfcc_root_object.cpp]: fixed

File: src/goto-instrument/contracts/dynamic-frames/dfcc_root_object.cpp
Comment (remi-delmas-3000) [reply src/goto-instrument/contracts/dynamic-frames/dfcc_root_object.cpp]: fixed doc

File: src/goto-instrument/contracts/dynamic-frames/dfcc_root_object.cpp
Comment (remi-delmas-3000) [reply src/goto-instrument/contracts/dynamic-frames/dfcc_root_object.cpp]: fixed

File: src/goto-instrument/contracts/dynamic-frames/dfcc_root_object.cpp
Comment (remi-delmas-3000) [reply src/goto-instrument/contracts/dynamic-frames/dfcc_root_object.cpp]: fixed

**Review by qinheping — APPROVED** [review]
LGTM

**Comment by feliperodri** [comment]
@tautschnig if you have the time, could you do another round of review on this PR?

File: src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.h
Comment (tautschnig) [reply src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.h]: Shouldn't this now be `assignst` as you suggest?

File: src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.h
Comment (tautschnig) [inline src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.h]: We usually use a `message_handlert`. But, also, you might get rid of it once the comments in the depended-on PR are addressed.

File: src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.cpp
Comment (tautschnig) [inline src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.cpp]: I had to read the remainder of the code to understand what was meant by "non-constant" here. I think this needs a bit more text.

File: src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.cpp
Comment (tautschnig) [inline src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.cpp:96]: What happens behind the scenes with a pointer passed to object-whole? Would it be the same as the `root_object` case above?

File: src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.cpp
Comment (tautschnig) [inline src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.cpp]: Just use `{{expr}}` or else `std::move(args)`.

File: src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.cpp
Comment (tautschnig) [inline src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.cpp]: Is this needed?

**Review by tautschnig — COMMENTED** [review]
Just minor cleanup, will be good to go once the depended-on PR is merged.

File: src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.h
Comment (remi-delmas-3000) [reply src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.h]: fixed.

File: src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.h
Comment (remi-delmas-3000) [reply src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.h]: fixed

File: src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.cpp
Comment (remi-delmas-3000) [reply src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.cpp]: added example. But this notion is actually implemented in `havoc_utils_is_constantt`

File: src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.cpp
Comment (remi-delmas-3000) [reply src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.cpp:96]: The meaning of  `object_whole(&expr)` or `object_whole(&root_object(expr))` is the same so we don't bother computing the root object. In the first case we compute the root_object to try and get rid of a dependency on loop locals or drop the target if the root object is still local to the loop. In the second case we have a non-constant expression that does not depend on loop locals, and we just widen it directly.

File: src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.cpp
Comment (remi-delmas-3000) [reply src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.cpp]: fixed

File: src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.cpp
Comment (remi-delmas-3000) [reply src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.cpp:96]: we have to use `object_whole` because this is the predicate that DFCC understands to specify byte ranges.

File: src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.cpp
Comment (remi-delmas-3000) [reply src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.cpp]: cleaned up

**Comment by remi-delmas-3000** [comment]
@tautschnig all suggestions implemented, could you please take a last look ?

**Review by tautschnig — APPROVED** [review]

**Pull request merged**
